File: mac/KhaytCore/Sources/KhaytCore/JS/price-agreements.js
'use strict';

/**
 * What a customer has agreed to pay for a thing, applied to the cart.
 *
 * A customer's record can carry a price list — "brackets: 12.50", "any
 * keychain: 8" — and when that customer is chosen for a job, each part whose
 * name contains one of those products takes the agreed figure. The match is
 * by name, case-insensitively, and the FIRST entry whose product the part's
 * name contains is the one that counts, even when its price is zero: a shop
 * that wrote a product down with no price has said "this one is not agreed",
 * and a later entry must not quietly stand in for it.
 *
 * ── WHERE THE FIGURE LANDS, AND WHY THAT IS SAID OUT LOUD ────────────────
 *
 * The agreed figure is written as the part's unit COST (`unitCost`, and
 * `baseCost` = cost × qty), which is what `renderer/wire-events.js` has always
 * done — and the job is then priced by the ordinary rule, cost plus the job's
 * margin. So a part agreed at 50 on a job quoted at 30% margin comes to 65
 * before discount. The shop sees that total on the form before saving and can
 * set the margin; nothing is hidden. But it means "price agreement" is, as
 * the book stores it, an agreed COST. Whether the rule ought to replace
 * cost-plus-margin for that part instead is a pricing question with a data
 * migration behind it (`lib/pricing.js` prices a cart, not a part), and this
 * module does not decide it — it makes both apps give the same answer, and
 * names the question in its test so it is not forgotten.
 *
 * PURE apart from the parts it is handed, which it mutates in place — the
 * cart in the Electron window is the array the form is drawn from.
 */
(function (global) {

  const listOf = (v) => (Array.isArray(v) ? v : []);

  /**
   * The agreement that covers a part with this name, or null.
   *
   * Null for a nameless part, for an entry with no product, and — see above —
   * when the first entry whose product the name contains has no price.
   */
  function find(priceList, name) {
    const wanted = String(name || '').toLowerCase();
    if (!wanted) return null;
    const entry = listOf(priceList).find(p => p && p.product
      && wanted.includes(String(p.product).toLowerCase()));
    return entry && +entry.price > 0 ? entry : null;
  }

  /**
   * Apply a customer's agreements to a cart. Returns how many parts took one,
   * so a host can say "price agreement applied" only when one was.
   */
  function apply(parts, priceList) {
    let applied = 0;
    for (const part of listOf(parts)) {
      const entry = find(priceList, part && part.name);
      if (!entry) continue;
      part.unitCost = +entry.price;
      part.baseCost = +entry.price * (+part.qty || 1);
      applied++;
    }
    return applied;
  }

  const api = { find, apply };
  if (typeof module !== 'undefined' && module.exports) module.exports = api;
  global.KhaytPriceAgreements = api;
})(typeof globalThis !== 'undefined' ? globalThis : this);

File: mac/KhaytCore/Sources/KhaytCore/PriceAgreements.swift
import Foundation

/// What a customer has agreed to pay for a thing — ported to Swift.
///
/// A customer's record can carry a price list — "brackets: 12.50", "any
/// keychain: 8" — and when that customer is chosen for a job, each part whose
/// NAME CONTAINS one of those products takes the agreed figure.
///
/// ── WHERE THE FIGURE LANDS, AND WHY IT MATTERS ────────────────────────────
///
/// On the part as `agreedPrice`, per unit, and nowhere else. The part's cost
/// stays what it cost, and the pricing rule charges the agreed figure for that
/// part INSTEAD of cost plus margin.
///
/// Until September 2026 the other app wrote the agreed figure into the COST,
/// so the job's margin went on top of it: a part agreed at 50 on a 30% job
/// billed 65, and the profit report then showed the part sold at cost. Both
/// were wrong, and the shop had to notice the total. An agreed price is a
/// PRICE, not a cost.
public enum PriceAgreements {

    /// The agreement covering a part with this name, or nil.
    ///
    /// ── THE FIRST MATCH WINS, EVEN WHEN IT HAS NO PRICE ───────────────────
    ///
    /// `find` takes the FIRST entry whose product the name contains, and only
    /// then asks whether that entry has a price. So a shop that wrote a
    /// product down with no price has said "this one is not agreed", and a
    /// later entry must not quietly stand in for it.
    ///
    /// Reading it as "the first entry with a price" would silently price a
    /// part the shop deliberately left open — which is the kind of difference
    /// nobody notices until an invoice is wrong.
    public static func find(in priceList: [JSONValue], name: JSONValue?) -> JSONValue? {
        let wanted = JSSemantics.truthy(name) ? JSSemantics.text(name).lowercased() : ""
        guard !wanted.isEmpty else { return nil }
        let entry = priceList.first { row in
            guard JSSemantics.truthy(row), case .object(let p) = row,
                  JSSemantics.truthy(p["product"]) else { return false }
            let product = JSSemantics.text(p["product"]).lowercased()
            // ── EVERY STRING CONTAINS THE EMPTY STRING ────────────────────
            //
            // `wanted.includes("")` is TRUE in JavaScript, and Swift's
            // `contains("")` is not — so an agreement whose product is
            // something truthy that STRINGIFIES to nothing, like `[]`, matches
            // every part and prices the whole cart at its figure. It is first
            // in the list, so it wins outright.
            //
            // That is a real trap in a real book: the products come from a
            // customer record somebody typed. Reproduced rather than tidied,
            // because tidying it would mean the two apps bill differently for
            // the same cart.
            if product.isEmpty { return true }
            return wanted.contains(product)
        }
        guard case .object(let found)? = entry,
              JSSemantics.number(found["price"]) > 0 else { return nil }
        return entry
    }

    /// The agreed price for a part with this name, or nil.
    public static func price(in priceList: [JSONValue], name: JSONValue?) -> Double? {
        guard case .object(let entry)? = find(in: priceList, name: name) else { return nil }
        return JSSemantics.number(entry["price"])
    }

    /// Apply a customer's agreements to a cart.
    ///
    /// A part with no agreement is left as it was EXCEPT that `agreedPrice` is
    /// removed — including one a previous customer left on it, because the
    /// cart now belongs to this customer. Returns the parts and how many took
    /// an agreement, so a host can say "price agreement applied" only when one
    /// was.
    public static func apply(to parts: [JSONValue],
                             priceList: [JSONValue]) -> (parts: [JSONValue], applied: Int) {
        var applied = 0
        let out = parts.map { row -> JSONValue in
            guard JSSemantics.truthy(row), case .object(var part) = row else { return row }
            guard let agreed = price(in: priceList, name: part["name"]) else {
                part["agreedPrice"] = nil        // `delete part.agreedPrice`
                return .object(part)
            }
            part["agreedPrice"] = .number(agreed)
            applied += 1
            return .object(part)
        }
        return (out, applied)
    }
}
